### ISSUE TITLE
Gradients through random sampling should be Null for MRG_RandomStreams distributions

### ISSUE THREAD

**Issue by lamblin** [Theano/Theano#4078]
The gradients wrt variables passed to methods of `MRG_RandomStreams` (for instance normal, binomial, ...) should be undefined / NaN (NullType), since this is not mathematically defined.

If users want to use some reparametrization trick, then they would have to, for instance, sample from a normal(0, 1), and then multiply by the variance and add the mean. In this case, it would become clear what gradients are expected.

The way of doing that would be to wrap the arguments in [`grad_undefined`](http://deeplearning.net/software/theano/library/gradient.html#theano.gradient.grad_undefined) before starting the chain of operations transforming the uniform sample into the sample of the required distribution.

**Comment by Sentient07** [comment]
@lamblin , I'd like to take this up ? Shall I go ahead ?

**Comment by lamblin** [comment]
I think we are currently lacking the time to keep up with your PRs (I think you have 4 at the moment), so maybe when we are done with some of the existing ones, you could start on this ticket?
Sorry for the slow response, but we have other duties as well...

**Comment by Sentient07** [comment]
@lamblin Yes, i understand :). Of that 4, i guess 3 are done and one of them is remaining. Review at your ease :smile:

**Comment by lamblin** [comment]
@SinaHonari This is your new ticket.

**Comment by lamblin** [comment]
So, the plan would be:
- create a new UndefinedGrad Op, inheriting from ViewOp, in theano/gradient.py. It would be similar to ZeroGrad or DisconnectedGrad, except that the grad() method would return grad_undefined(...) for each input.
- instantiate that Op for each of the inputs (bias, variance, ...) of the random functions that are the methods of the MRG_RandomStreams class (uniform, normal, multinomial, ...)

**Comment by lamblin** [comment]
Fixed by #5538.
